Big rewrite

SimonHeybrock opened the pull request:
This adds, as `scippnexus.v2`, the future version of ScippNexus. The trigger for this rewrite were significant performance issues with small files that contain thousands of groups and datasets. However, it also attempts to resolve some complicated logic of the old implementation, which did lead, e.g., to hard to understand stack traces in exceptions.

To use this, just try `import scippnexus.v2 as snx`. The basic API is the same as before.

Apart from internal design and significant performance improvements (not affecting large datasets), there are a number of changes:

- Fallback to loading as `DataGroup` is not better at preserving "partial" information, if assembly fails.
- Transformation chains are not executed and not resolved. We will provide post-processing functions for this.
- `NX*geometry` classes are loaded as `DataGroup` without processing. We will provide post-processing functions for grouping this by `detector_number`, etc.
- `NXevent_data` is not automatically grouped by `detector_number`. There is a post-processing function for this, but we will need to add some that operate recursively.
- Error-fields are "absorbed" into data fields. Accessing them independently is not possible any more. May change my mind about this again...

For review:

- Start with `base.py`. `class File` and its helper functions are mostly (but not completely) unchanged.
- Continue with `nxdata.py`.
- Look at the rest in any order.
- There is some unused code in `nxtransformations.py`, which I plan to refactor in a follow-up into a couple of user-facing functions for resolving and executing `depends_on`-chains.

Bugfixes:

- Fields with `shape=(0,)` are now correctly loaded as datetime if they should. Previously this always returned a time (offset).
- Fields with `shape=(0,)` are now correctly loaded as transformation if they should. Previously this always returned the plain value of the transformation.

jl-wynen commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+                        and values.dataset.shape == errors.dataset.shape):
+                    values.errors = errors.dataset
+                    del items[f'{name}{suffix}']
+        items = {k: v for k, v in items.items() if not k.startswith('cue_')}
```
Do this before identifying errors to reduce size of `items` early.

jl-wynen commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+                if sel.startswith('/'):
+                    return self.file[sel[1:]]
+                else:
+                    return self[sel.split('/')[0]][sel[sel.index('/') + 1:]]
```
```suggestion
                    grp, path = self.split('/', 1)
                    return self[grp][path]
```
?

SimonHeybrock replied:
Copilot agrees with you :D

jl-wynen replied:
See, it's not better than us as a team yet :-)

jl-wynen commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+
+def create_field(group: H5Group, name: str, data: DimensionedArray,
+                 **kwargs) -> H5Dataset:
+    if not isinstance(data, sc.Variable):
```
The function could be called with a data array as that satisfies `DimensionedArray`. This branch looks like it is meant for `np.ndarray`. But that doesn't satisfy the protocol.

SimonHeybrock replied:
Add np.ndarray to the type hint

jl-wynen replied:
My first, poorly worded, point remains. If `create_field` is called with a DataArray, it would hit this branch and try to call `group.create_dataset` with the data array.

SimonHeybrock replied:
This was fixed earlier, github did not pick it up due to move to a different file

jl-wynen reviewed (commented):
Only looked at `base.py` so far.

jl-wynen commented on src/scippnexus/v2/application_definitions/nxcansas/__init__.py:
```diff
@@ -0,0 +1,9 @@
+
+from .nxcansas import SASdata, SASentry, definitions
+
+__all__ = [definitions, SASentry, SASdata]
```
These should be strings.

jl-wynen commented on src/scippnexus/v2/nxsample.py:
```diff
@@ -0,0 +1,36 @@
+
+from .base import Field, Group, NXobject, ScippIndex, base_definitions
+
+_matrix_units = dict(zip(['orientation_matrix', 'ub_matrix'], ['one', '1/Angstrom']))
```
Write as a dict-literal to make it easier to see associations between names and units.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+    @staticmethod
+    def assemble_as_child(children: sc.DataGroup,
+                          detector_number: Optional[sc.Variable] = None) -> sc.Variable:
+        return sc.scalar(children)
```
This is actually unused now and should be removed.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+            if self.dataset.name.endswith('depends_on') and variable.ndim == 0:
+                variable.value = depends_on_to_relative_path(variable.value,
+                                                             self.dataset.parent.name)
+        elif variable.values.flags["C_CONTIGUOUS"]:
```
The `Eigen` storage order is different, iirc, that is why we have this.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+            starts = []
+            for name in self.attrs:
+                if (dt := _as_datetime(self.attrs[name])) is not None:
+                    starts.append(dt)
```
There is unfortunately no clear convention, they are typically named `start` or `offset`, if I remember correctly.

jl-wynen replied:
This seems risky. What if a tof dataset has a datetime attribute for some reason?

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+                variable = convert_time_to_datetime64(
+                    variable,
+                    start=starts[0],
+                    scaling_factor=self.attrs.get('scaling_factor'))
```
TOF does not have a `start` "datetime".

jl-wynen replied:
Is this guaranteed? This whole mechanism seems brittle. If there happens to be a single datetime attr, that gets used as an offset.

SimonHeybrock replied:
No it is not guaranteed. But if you store a datetime attr, isn't it clear that you mean absolute times? Or can you think of any other reason?

jl-wynen replied:
I guess things like runs start time, time of the first pulse, etc are stored in dedicated places and not randomly attached to tof datasets?

SimonHeybrock replied:
Yes, as far as I have observed.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+
+    @property
+    def name(self) -> str:
+        return self.dataset.name
```
I did not add caching unless I had performance benchmarks backing that up.

jl-wynen replied:
ok

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+        self._group = group
+        self._definitions = {} if definitions is None else definitions
+
+    @property
```
See above, may be added if we have benchmarks supporting that.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+        children, as well as defining how children will be read and assembled into the
+        result object when the group is indexed.
+        """
+        return self._definitions.get(self.attrs.get('NX_class'),
```
The key in `_definitions` is a string though?

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+    After loading we will generally not have the same root so absolute paths
+    cannot be resolved after loading."""
+    if depends_on.startswith('/'):
+        return os.path.relpath(depends_on, parent_path)
```
Good point, switching `os.path` -> `posixpath`.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+            time = time_and_timezone_offset[0]
+            if len(time_and_timezone_offset) == 1:
+                # No timezone, parse directly (scipp based on numpy)
+                return sc.datetime(f'{date}T{time}')
```
Not sure that is the same, doesn't that leave a final `Z`, if present?

jl-wynen replied:
No, this branch is for the case that there is no timezone offset.

SimonHeybrock replied:
But there may be a `Z`, with no offset, unless I misunderstand the code above?

jl-wynen replied:
But
```python
>>> re.split(r'Z|\+|-', '10Z')
['10', '']
```

So this branch will not get used in this case. It will only be used if there is no `Z`, `+`, or `-` at all.

SimonHeybrock replied:
Applied.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+    return unit.to_dict().get('powers') == {'s': 1}
+
+
+def _as_datetime(obj: Any):
```
There is a number of tests in `nexus_test.py`. Look for `timezone`, for example.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+    np.dtype('float32'): sc.DType.float32,
+    np.dtype('float64'): sc.DType.float64,
+    np.dtype('bool'): sc.DType.bool,
+}
```
Not sure, I think we will actually have to do better in the future (such as, support for dtypes, and validate overflows of uint64 when converting to int64).

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+
+
+def _dtype_fromdataset(dataset: H5Dataset) -> sc.DType:
+    return _scipp_dtype.get(dataset.dtype, sc.DType.string)
```
I have not encountered such a case. Does HDF5 support this?

jl-wynen replied:
Yes. They are plain bytes arrays and are loaded as `np.void` objects. I don't know if NeXus allows them or if they occur in practice. They are not really usable unless you can figure out their dtype.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+    sizes: Optional[Dict[str, int]] = None
+    dtype: Optional[sc.DType] = None
+    errors: Optional[H5Dataset] = None
+    _is_time: Optional[bool] = None
```
Because it is not meant for the user: It is just a flag overriding the logic. Things that do not have this set can also be a time. It used to be a constructor parameter, we can consider making it a method?

SimonHeybrock replied:
Ok, while refactoring it I actually discovered a small bug (the test was also wrong!), where it skips the conversion to datetime in the current mechanism because it returned early for `shape=(0,)`. Good "catch", thanks. 👍

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+    return _scipp_dtype.get(dataset.dtype, sc.DType.string)
+
+
+@dataclass
```
Isn't `__eq__` just comparing all fields?

jl-wynen replied:
I don't know if it only compares dataclass fields or all attributes. Note that `cached_property` stores the cached value as an instance attribute. So

- if `__eq__` and `__hash__` only look at dataclass fields, they ignore caches which is probably what we want.
- if they look at all instance attributes, then an instance that has cached values will compare not equal to one that has no cached values.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+        return self.parent.file
+
+    def _load_variances(self, var, index):
+        stddevs = sc.empty(dims=var.dims,
```
I used scipp since I wanted multi-threading for the `pow` call below.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+        var.variances = np.broadcast_to(sc.pow(stddevs, sc.scalar(2)).values,
+                                        shape=var.shape)
+
+    def __getitem__(self, select) -> Union[Any, sc.Variable]:
```
It is copied from the old implementation. Can do, but maybe in a follow up?

jl-wynen replied:
Ok. But note https://github.com/scipp/scippnexus/pull/117#discussion_r1154437855

SimonHeybrock commented on src/scippnexus/v2/base.py:
It is in the same file for the easy of type hints. Since the all reference each other. Is there a nice way?

jl-wynen replied:
If it is only about type checking, you can use https://docs.python.org/3/library/typing.html?highlight=type_checking#typing.TYPE_CHECKING to prevent import cycles. This is not super nice but I'd say this here is a valid use case.

SimonHeybrock replied:
I think I will move `Field` to a new file. Keeping `NXobject` and `Group` in the same makes more sense, as they have tighter coupling.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+        of any other field. For example, field attributes may define which fields are
+        axes, and dim labels of other fields can be defined by the names of the axes.
+        """
+        _ = self._nexus
```
Added docstring to `NXobject.__init__`.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+class NXgeometry(NXobject):
+
+    def __init__(self, attrs: Dict[str, Any], children: Dict[str, Union[Field, Group]]):
+        super().__init__(attrs=attrs, children=children)
```
Removed NXgeometry from `base.py`, as it is not longer necessary.

SimonHeybrock commented:
Apart from addressing your comments, I pushed another small change the switches out the `cached_property` for manual lazy init for the `Group._children` and  `Group._nexus` properties. This should make it easier to see through the logic. I kept `cached_property` for the other methods/properties that "just" return something simple that is computed.

jl-wynen commented:
> 

I honestly don't see the benefit. Was the logic unclear before?

jl-wynen commented on src/scippnexus/v2/nxdata.py:
```diff
@@ -0,0 +1,357 @@
+            if name in self._aux_signals:
+                return _guess_dims(group_dims, self._signal.dataset.shape,
+                                   field.dataset)
+            # Latest way of defining dims
```
So should this one be tried first to override legacy methods?

SimonHeybrock replied:
You mean before the `_aux_signals` check? Yes, that may make sense. The docs are not clear of the `_indices` attributes would be used for auxiliary signals (it probably assumes the have the same shape as the data), but I can't see anything wrong with using the property.

jl-wynen commented on src/scippnexus/v2/nxdata.py:
```diff
@@ -0,0 +1,357 @@
+                 dg: sc.DataGroup) -> Union[sc.DataGroup, sc.DataArray, sc.Dataset]:
+        if not self._valid:
+            raise NexusStructureError("Could not determine signal field or dimensions.")
+        aux = {name: dg.pop(name) for name in self._aux_signals}
```
This modifies the argument `dg`. But below you made a copy fog `coords` before modifying it. Seems to me like you should not use `dg` here directly either.

SimonHeybrock commented:
> > 
> 
> I honestly don't see the benefit. Was the logic unclear before?

In particular the `_children` felt odd: We write to the return value, but it is not simply returning an actual field of the class instance. Rather, it is returning something that it constructed in the property. We basically relied on `cached_property` to do this exactly once, and this may be unexpected to the reader.

SimonHeybrock commented on src/scippnexus/v2/nxdata.py:
```diff
@@ -0,0 +1,357 @@
+        super().__init__(attrs=attrs, children=children, fallback_signal_name='data')
+
+
+def _group_events(*,
```
They somewhat depend on `NXdetector` that is why they are here. I could move the lower-level ones.

jl-wynen replied:
I'd move all or none of them to avoid exposing implementation details. So leave them here.

SimonHeybrock commented on src/scippnexus/v2/nxevent_data.py:
```diff
@@ -0,0 +1,158 @@
+            return (_event_dimension, )
+        return None
+
+    def read_children(self, obj: Group, select: ScippIndex) -> sc.DataGroup:
```
Shorter than before ;)

SimonHeybrock replied:
Extracted the index setup bit.

jl-wynen replied:
One thing that took me a moment to understand is the handling of `event_id`. It is computed somewhat early but only used in the very last statement before return with a repeated check for `event_id is not None`. Can you move those two pieces of code together to clarify?

SimonHeybrock replied:
Done, and removed the specific dtype expection.

SimonHeybrock commented on src/scippnexus/v2/base.py:
```diff
@@ -0,0 +1,673 @@
+                errors = items[f'{name}{suffix}']
+                if (values.unit == errors.unit
+                        and values.dataset.shape == errors.dataset.shape):
+                    values.errors = errors.dataset
```
Should be fixed now.

jl-wynen commented on tests/nxtransformations_test.py:
```diff
@@ -2,90 +2,169 @@
 @pytest.fixture()
-def nxroot():
-    """Yield NXroot containing a single NXentry named 'entry'"""
+def h5root(request):
```
Unused param `request`?

SimonHeybrock commented:
Should be ready for another review @jl-wynen.

jl-wynen commented on src/scippnexus/v2/nxdata.py:
```diff
@@ -99,15 +77,8 @@ def _get_group_dims() -> Optional[Tuple[str, ...]]:
-            named_axes = fallback_dims
-        else:
-            named_axes = ()
+        self._group_dims = group_dims
```
Why do you store the group dims as an instance attribute now?
Is it because of `_get_dims`? In that case, can't you just pass the group dims as an argument?

jl-wynen commented on src/scippnexus/v2/nxdata.py:
```diff
@@ -116,39 +87,11 @@ def _get_group_dims() -> Optional[Tuple[str, ...]]:
 
         dims = np.array(group_dims)
-        dims_from_indices = {
+        self._dims_from_indices = {
```
Same question as `self._group_dims`.

jl-wynen commented on tests/nx2_test.py:
```diff
@@ -0,0 +1,209 @@
+    data['time'] = np.arange(5.0)
+    data['time'].attrs['units'] = 's'
+    data['time_errors'] = np.arange(5.0)
+    data['time_errors'].attrs['units'] = 's'
```
Does it make sense to add a test where the error unit differs from the values unit? This would test that snx detects the mismatch.

SimonHeybrock replied:
Adding one in `nexus_test.py`

jl-wynen commented on tests/nx2_test.py:
Does this file test new behaviour? Are you planning to merge this into the other test files?

SimonHeybrock replied:
Split, and added `nxevent_data_test.py`

jl-wynen reviewed (commented):
I have gone through all files now.

SimonHeybrock commented:
Tracking next steps in #120.

SimonHeybrock commented:
@jl-wynen I fixed another small issue, and pushed some extra horrible code to deal with event-data fields embedded in `NXdetector`, to somewhat restore the ability to load SNS files. I am not planning further changes in this PR, so this is ready for a "final" look.
